### PR TITLE
storage: assert on update_read_capabilities

### PR DESCRIPTION
**Pull request by aljoscha**
For running CI.

### CODE REVIEW

**Comment by aljoscha** [comment]
This is failing because compute controller is trying to acquire holds that are not beyond the current frontier of `read_capabilities`:

```
thread 'coordinator' panicked at 'update ChangeBatch { updates: [(1676396769846, 1), (1676396798000, -1)], clean: 2 } for collection s307 is trying to install read capabilities before the current frontier of read capabilities, read capabilities before applying: MutableAntichain { dirty: 0, updates: [(1676396798000, 4), (1676396810000, 1)], frontier: [1676396798000], changes: ChangeBatch { updates: [], clean: 0 } }', /home/aljoscha/Dev/work/materialize/src/storage-client/src/controller.rs:1571:25
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at ./rustc/fc594f15669680fa70d255faec3ca3fb507c3405/library/core/src/panicking.rs:64:14
   2: <mz_storage_client::controller::Controller<mz_repr::timestamp::Timestamp> as mz_storage_client::controller::StorageController>::update_read_capabilities
   3: <mz_compute_client::controller::instance::ActiveInstance<mz_repr::timestamp::Timestamp>>::update_write_frontiers
             at ./home/aljoscha/Dev/work/materialize/src/compute-client/src/controller/instance.rs:893:13
   4: <mz_compute_client::controller::instance::ActiveInstance<mz_repr::timestamp::Timestamp>>::handle_frontier_uppers
             at ./home/aljoscha/Dev/work/materialize/src/compute-client/src/controller/instance.rs:1087:9
   5: <mz_compute_client::controller::instance::ActiveInstance<mz_repr::timestamp::Timestamp>>::handle_response
             at ./home/aljoscha/Dev/work/materialize/src/compute-client/src/controller/instance.rs:1042:17
   6: <mz_compute_client::controller::ActiveComputeController<mz_repr::timestamp::Timestamp>>::process
             at ./home/aljoscha/Dev/work/materialize/src/compute-client/src/controller.rs:603:24
   7: <mz_controller::Controller>::process::{closure#0}
             at ./home/aljoscha/Dev/work/materialize/src/controller/src/lib.rs:298:32
   8: <mz_adapter::coord::Coordinator>::handle_message::{closure#0}
             at ./home/aljoscha/Dev/work/materialize/src/adapter/src/coord/message_handler.rs:49:21
   9: <mz_adapter::coord::Coordinator>::serve::{closure#0}
             at ./home/aljoscha/Dev/work/materialize/src/adapter/src/coord.rs:1179:37
  10: <tokio::runtime::park::CachedParkThread>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>::{closure#0}
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/park.rs:283:63
  11: tokio::runtime::coop::with_budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<<mz_adapter::coord::Coordinator>::serve::{closure#0}>::{closure#0}>
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/coop.rs:102:5
  12: tokio::runtime::coop::budget::<core::task::poll::Poll<()>, <tokio::runtime::park::CachedParkThread>::block_on<<mz_adapter::coord::Coordinator>::serve::{closure#0}>::{closure#0}>
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/coop.rs:68:5
  13: <tokio::runtime::park::CachedParkThread>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/park.rs:283:31
  14: <tokio::runtime::context::BlockingRegionGuard>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/context.rs:315:13
  15: <tokio::runtime::handle::Handle>::block_on::<<mz_adapter::coord::Coordinator>::serve::{closure#0}>
             at ./cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/handle.rs:264:9
  16: mz_adapter::coord::serve::{closure#0}::{closure#1}
             at ./home/aljoscha/Dev/work/materialize/src/adapter/src/coord.rs:1356:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Observed when running:
```
bin/mzcompose --find platform-checks run default --scenario=RestartEnvironmentdClusterdStorage
```

File: src/storage-client/src/controller.rs [hunk 1]
```diff
@@ -1556,6 +1556,29 @@ where
...
+                        assert!(
+                            current_read_capabilities.less_equal(time),
+                            "update {:?} for collection {key} is trying to \
+                            install read capabilities before the current \
+                            frontier of read capabilities, read capabilities before applying: {:?}",
+                            update,
+                            collection.read_capabilities
+                        );
+                    }
+                }
```
Comment (teskje) [inline src/storage-client/src/controller.rs:1580]: Would you mind adding the same asserts to the `update_read_capabilities` method in the compute controller? The code is similar enough that you should be able to copy-paste most of it.

**Comment by aljoscha** [comment]
@teskje Should we now merge this?

**Review by teskje — APPROVED** [review]

**Comment by teskje** [comment]
Let's do it!

**Pull request merged**